app/lifecycle/lifecycle.go

```diff
@@ -25,26 +28,61 @@ func Run() {
 	if err != nil {
 		log.Fatalf("Failed to start: %s", err)
 	}
+	callbacks := t.GetCallbacks()
 
-	var done <-chan int
+	go func() {
+		slog.Debug("XXX starting callback handler")
+		for {
+			select {
+			case <-callbacks.Quit:
+				slog.Debug("QUIT called")
+				t.Quit()
+			case <-callbacks.Update:
+				slog.Debug("XXX about to call DoUpgrade")
+				err := DoUpgrade()
+				if err != nil {
+					slog.Debug(fmt.Sprintf("DoUpgrade FAILED: %s", err))
+				}
+			case <-callbacks.ShowLogs:
+				slog.Debug("SHOW LOGS CALLED")
+				ShowLogs()
+			case <-callbacks.DoFirstUse:
+				slog.Debug("XXX Would be popping up powershell with a custom profile")
+			}
+		}
+	}()
+
+	// Are we first use?
+	if store.GetFirstTimeRun() {
+		err = t.DisplayFirstUseNotification()
+		if err != nil {
+			slog.Debug(fmt.Sprintf("XXX failed to display first use notification %v", err))
+		}
+	}
+
+	var done chan int
 	if IsServerRunning(ctx) {
-		log.Printf("XXX detected server already running")
+		slog.Debug("XXX detected server already running")
```

Comment:
I think we could probably remove the check for `IsServerRunning` and just keep trying to restart it every 5s or so

Reply:
I'll pick this up on my next pass of refinements.